### ISSUE TITLE
`operations/check` REST API

### ISSUE THREAD

**Issue by hongbo-miao** [rclone/rclone#7015]
#### The associated forum post URL from `https://forum.rclone.org`

Originally asked at https://forum.rclone.org/t/what-is-corresponding-rest-api-for-rclone-check/38283/2

#### What is your current rclone version (output from `rclone version`)?

```shell
rclone v1.62.2
- os/version: darwin 13.3.1 (64 bit)
- os/kernel: 22.4.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.20.2
- go/linking: dynamic
- go/tags: none
```

#### What problem are you are trying to solve?

I started a server at
```
rclone rcd --rc-web-gui --rc-addr=:5572 --rc-user=admin --rc-pass=passw0rd
```

I am trying to find the corresponding REST API for this command

```
rclone check --one-way --use-json-log --include="*.tdms" /tmp/motor hm-s3:hongbomiao-bucket/iot/motor
```

I was expecting something like
```
POST http://localhost:5572/operations/check
```

However, operations/check does not exist at [Remote Control / API](https://rclone.org/rc/)

#### How do you think rclone should be changed to solve that?

It would be great to add `operations/check` at [Remote Control / API](https://rclone.org/rc/). 😃

**Comment by Jwink3101** [comment]
Just to add, a `cryptcheck`, especially when you can specify the two files, potentially with different names, would be very helpful

**Comment by Viktor-Yakovchuk** [comment]
I want to implement this.

**Comment by ncw** [comment]
@Viktor-Yakovchuk I already did this sorry!

@Hongbo-Miao please give this a try and let me know what you think of the interface.

[v1.63.0-beta.7054.3d0a2336b.fix-7015-rc-check](https://beta.rclone.org/branch/fix-7015-rc-check/v1.63.0-beta.7054.3d0a2336b.fix-7015-rc-check/) on branch [fix-7015-rc-check](https://github.com/rclone/rclone/tree/fix-7015-rc-check) (uploaded in 15-30 mins)

I haven't added tests for this yet.

Here are the docs

### operations/check: check the source and destination are the same {#operations-check}

Checks the files in the source and destination match.  It compares
sizes and hashes and logs a report of files that don't
match.  It doesn't alter the source or destination.

This takes the following parameters:

- srcFs - a remote name string e.g. "drive:" for the source, "/" for local filesystem
- dstFs - a remote name string e.g. "drive2:" for the destination, "/" for local filesystem
- download - check by downloading rather than with hash
- checkFileHash - treat checkFileFs:checkFileRemote as a SUM file with hashes of given type
- checkFileFs - treat checkFileFs:checkFileRemote as a SUM file with hashes of given type
- checkFileRemote - treat checkFileFs:checkFileRemote as a SUM file with hashes of given type
- oneWay -  check one way only, source files must exist on remote
- combined - make a combined report of changes (default false)
- missingOnSrc - report all files missing from the source (default true)
- missingOnDst - report all files missing from the destination (default true)
- match - report all matching files (default false)
- differ - report all non-matching files (default true)
- error - report all files with errors (hashing or reading) (default true)

If you supply the download flag, it will download the data from
both remotes and check them against each other on the fly.  This can
be useful for remotes that don't support hashes or if you really want
to check all the data.

If you supply the size-only global flag, it will only compare the sizes not
the hashes as well.  Use this for a quick check.

If you supply the checkfile HASH flag with a valid hash name,
the srcFs:srcRemote must point to a text file in the SUM format.

Returns:

- success - true if no error, false otherwise
- status - textual summary of check, OK or text string
- hashType - hash used in check, may be missing
- combined - array of strings of combined report of changes
- missingOnSrc - array of strings of all files missing from the source
- missingOnDst - array of strings of all files missing from the destination
- match - array of strings of all matching files
- differ - array of strings of all non-matching files
- error - array of strings of all files with errors (hashing or reading)

**Authentication is required for this call.**

**Comment by Hongbo-Miao** [comment]
Thanks @ncw ! I didn't find the info regarding to `--include="*.tdms"` in this REST API.

**Comment by ncw** [comment]
If you are going to use `curl` with JSON input then you need to supply the [flag  -H "Content-Type: application/json"](https://rclone.org/rc/#using-post-with-a-json-blob), eg

    curl -H "Content-Type: application/json" -X POST -d '{"potato":2,"sausage":1}' http://localhost:5572/rc/noop

**Comment by ncw** [comment]
> And, I don't know how to add `--include="*.tdms"` in this REST API.

`--include` is a global filter flag. See [the docs](https://rclone.org/rc/#setting-filter-flags-with-filter) for how to set that,eg

    "_filter":{"MaxSize":"1M", "IncludeRule":["a","b"], "MaxAge":"42s"}

**Comment by ncw** [comment]
I rebased this on master.

[v1.65.0-beta.7382.3612d7ea9.fix-7015-rc-check](https://beta.rclone.org/branch/fix-7015-rc-check/v1.65.0-beta.7382.3612d7ea9.fix-7015-rc-check/) on branch [fix-7015-rc-check](https://github.com/rclone/rclone/tree/fix-7015-rc-check) (uploaded in 15-30 mins)

Testing appreciated!

**Comment by ncw** [comment]
Here is what I hope is the final iteration with tests and docs!

[v1.65.0-beta.7390.98bf5cdf7.fix-7015-rc-check](https://beta.rclone.org/branch/fix-7015-rc-check/v1.65.0-beta.7390.98bf5cdf7.fix-7015-rc-check/) on branch [fix-7015-rc-check](https://github.com/rclone/rclone/tree/fix-7015-rc-check) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.65

**Comment by hongbo-miao** [comment]
Thanks @ncw ! ☺️
